libopenage/curve/queue.h

```diff
@@ -183,24 +184,74 @@ class Queue : public event::EventEntity {
 	 */
 	container_t container;
 
-	iterator last_front;
+	/**
+     * The time of the last access to the queue.
+     */
+	time::time_t last_pop;
 };
 
 
 template <typename T>
-const T &Queue<T>::front(const time::time_t &t) const {
-	return this->begin(t).value();
+const T &Queue<T>::front(const time::time_t &time) const {
+	if (this->empty(time)) [[unlikely]] {
+		throw Error{MSG(err) << "Tried accessing front at "
+		                     << time << " but queue is empty."};
+	}
+
+	// search for the first element that is after the given time
+	auto it = this->container.begin();
+	++it;
+	while (it != this->container.end() and it->time() <= time) {
+		++it;
+	}
+
+	// get the last element before the given time
+	--it;
+
+	return it->value;
 }
 
 template <class T>
-bool Queue<T>::empty(const time::time_t &time) const {
-	return this->last_front == this->begin(time).get_base();
+const T Queue<T>::pop_front(const time::time_t &time) {
+	if (this->empty(time)) [[unlikely]] {
+		throw Error{MSG(err) << "Tried accessing front at "
+		                     << time << " but queue is empty."};
+	}
+
+	// search for the first element that is after the given time
+	auto it = this->container.begin();
+	++it;
+	while (it->time() <= time and it != this->container.end()) {
```

Comment:
I swapped the search order around, so we start from the end of the queue. Then we probably don't even need an index structure, since in normal gameplay, the front element is more likely to be closer to the end of the container due to how we insert elements:

    for (auto it = this->container.begin(); it != this->container.end(); ++it) {
		if (time < it->time()) {
			insertion_point = this->container.insert(it, queue_wrapper(time, e));
			break;
		}
	}
	if (insertion_point == this->container.end()) {
		insertion_point = this->container.insert(this->container.end(),
		                                         queue_wrapper(time, e));
	}